bip-vaults.mediawiki

```diff
@@ -0,0 +1,756 @@
+
+
+==== Check for recovery ====
+
+A check is performed to determine if this input is being spent within the context of
+a recovery transaction, exactly as in [[#check-for-recovery|the <code>OP_VAULT</code> evaluation described above]].
+
+
+==== <code>OP_UNVAULT</code> evaluation for recovery spend ====
+
+This is identical to the [[#op_vault-evaluation-for-recovery-spend|<code>OP_VAULT</code> case described above]].
+
+
+==== <code>OP_UNVAULT</code> evaluation for withdrawal ====
+
+When spending an <code>OP_UNVAULT</code> input into a withdrawal target, no witness stack is required.
+
+* <code><spend-delay></code> is used to check whether the withdrawal of the input has matured.
+** If the input's relative timelock check, as described in [https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki BIP-0112] (using this value as "the top item on the stack") fails, the script MUST fail and terminate immediately.
+*** The same <code>CheckSequence(<spend-delay>)</code> code path is used as for [https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki BIP-0112].
+
+* The transaction outputs are then checked to verify that the withdrawal outputs are as expected.
+** If <code>target_outputs_hash(spending_tx.vout) != <target-output-hash></code> per the algorithm defined below, the script MUST fail and terminate immediately.
+
+<source lang="python">
+def target_outputs_hash(vout: [CTxOut]) -> bytes:
+    return hash256(b"".join(serialize_txout(out) for out in vout))
+
+def serialize_txout(txo: CTxOut) -> bytes:
+    spk: bytes = txo.scriptPubKey
+    return struct.pack("<q", txo.nValue) + ser_compact_size(len(spk)) + spk
+</source>
+
+If the above conditions do not fail, a single true value (<code>0x01</code>) is pushed to the stack.
+
+== Policy changes ==
+
+In order to prevent possible pinning attacks, recovery transactions must be replaceable.
+
+* When validating an <code>OP_VAULT</code>/<code>OP_UNVAULT</code> input being spent towards a recovery, the script must FAIL (by policy, not consensus) and terminate immediately if neither<ref>'''Why are recovery transactions required to be replaceable?''' In the case of unauthorized recoveries, an attacker may attempt to pin recovery transactions by broadcasting a "rebundled" version with a low fee rate. Vault owners must be able to overcome this with replacement. In the case of authorized recovery, if an attacker steals the recovery authorization key, the attacker may try to pin the recovery transaction during theft. Requiring replaceability ensures that the owner can always raise the fee rate of the recovery transaction, even if they are RBF rule #3 griefed in the process.</ref>
+*# the input is marked as opt-in replaceable by having an nSequence number less than <code>0xffffffff - 1</code>, per [https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki BIP-0125], nor
+*# the version of the recovery transaction has an nVersion equal to 3.
```

Comment:
This is going to be misinterpreted in its current form. Perhaps this would be easier to human-parse?

```suggestion
* When validating an <code>OP_VAULT</code>/<code>OP_UNVAULT</code> input being spent towards a recovery, the script must FAIL (by policy, not consensus) and terminate immediately if both<ref>'''Why are recovery transactions required to be replaceable?''' In the case of unauthorized recoveries, an attacker may attempt to pin recovery transactions by broadcasting a "rebundled" version with a low fee rate. Vault owners must be able to overcome this with replacement. In the case of authorized recovery, if an attacker steals the recovery authorization key, the attacker may try to pin the recovery transaction during theft. Requiring replaceability ensures that the owner can always raise the fee rate of the recovery transaction, even if they are RBF rule #3 griefed in the process.</ref>
*# the input is not marked as opt-in replaceable by having an nSequence number less than <code>0xffffffff - 1</code>, per [https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki BIP-0125], and
*# the version of the recovery transaction has an nVersion other than 3.
```